scipy/signal/tests/test_dltisys.py

```diff
@@ -234,6 +234,39 @@ def test_dlsim_simple2d(self):
         assert_array_equal(yout, expected)
         assert_array_equal(xout, expected)
 
+    def test_dlsim_dtypes(self):
+        # test proper handling of output dtypes, including complex values
+                
+        cases = [
+            ('f4','f4','f4','f4','f4'),
+            ('c8','f4','f4','f4','f4'),
+            ('f4','f8',float,'f4','f4'),
+            (complex,'f4','c16','f4','f4'),
+            ('c8','f4','f4','f8','f8'),
+            (int,int,int,int,int),
+        ]
+
+        t = np.arange(10)
+        u0 = np.zeros(10)
+        A0 = np.zeros((4,4))
+        B0 = np.zeros((4,1))
+        C0 = np.zeros((1,4))
+        D0 = np.zeros(1)
+        x0 = None
+    
+        for case in cases:
+            u = u0.astype(case[0])
+            A = A0.astype(case[1])
+            B = B0.astype(case[2])
+            C = C0.astype(case[3])
+            D = D0.astype(case[4])
+            system = dlti(A, B, C, D)
+            _, yout, xout = dlsim(system, u, t, x0)
+
+            out_dtype = np.result_type(*(np.dtype(c) for c in case))
+            assert yout.dtype == out_dtype
+            assert xout.dtype == out_dtype
```

Comment:
Since no old tests were changed and only a new test was added, my initial intuition is that this doesn't seem super risky, but core dev comments have me a bit cautious here:

> We can't just make dlsim or dimpulse accept complex data. We have to make the whole module work with complex coefficient models. Otherwise it will be a mess for everyone.

One other thing on my mind (as a non-expert in `scipy.signal`)--if we add support for complex-valued types to a function, should we not check for correctness of the output beyond just the dtype pass-through? Or is it 100 % clear that supporting various exotic complex numbers will always work if real-type inputs give the right answer? My intuition is that we may want to spot check for correctness (not just dtype) with some complex inputs as well? 

For now though, I think I'll have to bump the release milestone to `1.13.0`, since `1.12.0` branching is pretty soon. That said, `1.13.0` may be coming out pretty soon to support NumPy `2.0.0`, so there may be an opportunity to at least move part of this forward. Might be good to have another `signal` regular chime in to the discussion.